pkg/puller/puller.go

```diff
@@ -148,23 +139,8 @@ func (p *Puller) manage() {
 				if _, ok := bp[peerAddr.String()]; ok {
 					delete(peersDisconnected, peerAddr.String())
 				}
-				syncing := len(bp)
-				if po < depth {
-					// outside of depth, sync peerPO bin only
-					if _, ok := bp[peerAddr.String()]; !ok {
-						if syncing < p.shallowBinPeers {
-							// peer not syncing yet and we still need more peers in this bin
-							bp[peerAddr.String()] = newSyncPeer(peerAddr, p.bins)
-							peerEntry := peer{addr: peerAddr, po: po}
-							peersToSync = append(peersToSync, peerEntry)
-						}
-					} else {
-						// already syncing, recalc
-						peerEntry := peer{addr: peerAddr, po: po}
-						peersToRecalc = append(peersToRecalc, peerEntry)
-					}
-				} else {
-					// within depth, sync everything >= depth
+				if po >= depth {
+					// within depth, sync everything
```

Comment:
As much as I currently dislike pullsync and how it pulls chunks into my node's localstore that have no business being there (well outside my neighborhood), I believe this change will result in leaving some chunks destined to never arrive in their desired neighborhood.  Consider the following situation:

An uploading node with a pusher queue drops network connectivity for some reason.  All peers are dropped due to this network disruption.  When the network event is past, the node will start establishing new peer connections.  As soon as a single connection is established, the pusher will flush the queue entirely via the first connected peer.  And if the node itself is closer to the chunk than that peer, then those chunks will be stored locally (topology.ErrWantSelf from pushToClosest) and will not be pushed again (see #1504 )

Without the puller rummaging around in nodes outside its own neighborhood, these misplaced chunks will never move outside the node that invalidly adopted them and will therefore be very unlikely to be discovered by any retrieval attempt which would be routed to those chunks' preferred neighborhood nodes.

I wish I could offer a fix for this, but I'm still wrapping my head around the issues with pullsync in general and working on formulating an issue and a (better) PR to correct those issues.

Reply:
@ldeffenb thanks we will keep this in mind. We're making improvements across the board and pushsync will also undergo more improvements soon. Thanks for the feedback :+1:

Reply:
I just tried this patch on my mutant 0.5.3 node and it is lacking one thing.  As the node starts up, the neighborhood is quite low and pullers are started for peers that will end up outside the neighborhood once a full complement of peers has been acquired and the neighborhood has increased.

I believe there needs to be an else clause to cancel peers that are now outside the neighborhood but already started syncing while the neighborhood of peers was being established.

I know this because I have debug logs showing me the peers going through SyncInterval and the first ones to appear are well out of my leading bits range and they continue to operate even as the connected peers is increasing the neighborhood bin index.  I added the following else to force a recalc on syncing peers that are no longer within the neighborhood depth.

				} else {
					if _, ok := bp[peerAddr.String()]; ok {	// Force a recalc on syncing peers no longer within depth
						// already syncing, recalc
						peerEntry := peer{addr: peerAddr, po: po}
						peersToRecalc = append(peersToRecalc, peerEntry)
					}
				}

But this is not enough because (at least my) recalcPeer routine still tries to keep some syncing for peers outside the current depth.  So I made sure that all of the bins for peers "outside of depth" end up in "dontWant" so that they will all be cancelled.

Reply:
@ldeffenb you will not be pulling for farther away peers. Once kademlia depth changes, puller also updates and cancels all streams outside of depth